servicetalk-http-api/src/main/java/io/servicetalk/http/api/Uri3986.java

```diff
@@ -169,7 +173,7 @@ final class Uri3986 implements Uri {
                 if (i == 0) {
                     throw new IllegalArgumentException("Invalid URI format: no scheme before colon (':')");
                 }
-                parsedScheme = uri.substring(0, i);
+                parsedScheme = getLowerCaseScheme(uri.substring(0, i));
```

Comment:
If you are using `regionMatches`, there is an opportunity to skip allocation of a `substring` for the happy path if you pass `uri` and `i` as separate arguments.

Reply:
Indeed!
Additionally, I think we can reduce the allocation of the substring:

```java
    private static String getLowerCaseScheme(String uri, int endIndex) {
        if (endIndex == 4 && uri.regionMatches(true, 0, HTTP_SCHEME, 0, 4)) {
            return HTTP_SCHEME;
        } else if (endIndex == 5 && uri.regionMatches(true, 0, HTTPS_SCHEME, 0, 5)) {
            return HTTPS_SCHEME;
        }
        final char[] scheme = new char[endIndex];
        for (int i = 0; i < scheme.length; ++i) {
            scheme[i] = Character.toLowerCase(uri.charAt(i));
        }
        return new String(scheme);
    }
```

What do you think it?

Reply:
Taking into account that this is an unlikely fallback, I think `substring` followed by `toLoweCase` will be more readable and likely more efficient. `new String` does yet another allocation and attempts to do a compact string, while `substring` already knows if the original `uri` is compact or not and simply copies a subrange. Then `toLowerCase` has a non-allocating path if everything is already in lowercase.